engine/consensus/matching/engine.go

```diff
@@ -908,15 +930,97 @@ func (e *Engine) requestPending() error {
 		blockID := header.ID()
 		if _, ok := knownResultForBlock[blockID]; !ok {
 			missingBlocksOrderedByHeight = append(missingBlocksOrderedByHeight, blockID)
+			continue
 		}
+
 	}
 
 	// request missing execution results, if sealed height is low enough
 	log.Info().
 		Int("finalized_blocks_without_result", len(missingBlocksOrderedByHeight)).
 		Msg("requesting receipts")
 	for _, blockID := range missingBlocksOrderedByHeight {
-		e.requester.EntityByID(blockID, filter.Any)
+		e.receiptRequester.EntityByID(blockID, filter.Any)
+	}
+
+	return nil
+}
+
+// requestPendingApprovals requests approvals for chunks that haven't collected
+// enough approvals. When the number of unsealed finalized blocks exceeds the
+// threshold, we go through the entire mempool of incorporated-results, which
+// haven't yet been sealed, and check which chunks need more approvals.
+func (e *Engine) requestPendingApprovals() error {
+	// last sealed block
+	sealed, err := e.state.Sealed().Head()
+	if err != nil {
+		return fmt.Errorf("could not get sealed height: %w", err)
+	}
+
+	// last finalized block
+	final, err := e.state.Final().Head()
+	if err != nil {
+		return fmt.Errorf("could not get finalized height: %w", err)
+	}
+
+	// only request if number of unsealed finalized blocks exceeds the threshold
+	log := e.log.With().
+		Uint64("finalized_height", final.Height).
+		Uint64("sealed_height", sealed.Height).
+		Uint("sealing_threshold", e.sealingThreshold).
+		Logger()
+	if uint(final.Height-sealed.Height) < e.sealingThreshold {
+		log.Debug().Msg("skip requesting approvals as number of unsealed finalized blocks is below threshold")
+		return nil
+	}
+
+	log.Info().Msg("requesting approvals")
+	for _, r := range e.incorporatedResults.All() {
```

Comment:
At the moment we already rely on `clearPools` to remove incorporated-results pertaining to sealed blocks from the `incorporatedResults` mempool. So, assuming that `clearPools` works correctly, `requestPendingApprovals` does not repeatedly make requests for blocks that are already sealed. 

In light of your second bullet point, we could update `clearPools` to also remove items pertaining to blocks that conflict with an existing finalized block.

I am not sure I understand your third bullet point.

I believe we should invest some time in rethinking the `clearPools` function, as @AlexHentschel had already expressed some concerns about it. It is a very important piece of logic that also affects the efficiency of the builder.

Reply:
Added this extra check to skip requesting approvals for results that correspond to sealed or unfinalized blocks:

https://github.com/onflow/flow-go/blob/4b91a746a6fa80dda2ad3d0d18d3c5f8bb657638/engine/consensus/matching/engine.go#L1005-L1009

I did not implement your suggestion of caching the incorporated-result's block height because I think the storage layer already does it's own caching.